<a href="https://colab.research.google.com/github/Faizall16/Assosiation-Rule-with-Apyori-Algorithm/blob/main/Assosiation_Rule_with_Apyori_Algorithm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Menganalisa Pembelian Barang Menggunakan Algoritma Apriori

Teman saya yang bernama Joko ingin membuat acara diskon besar-besaran pada toko online nya. Namun, Ia bingung dengan barang apa yang sering dibeli para pembelinya. Disini saya memberitahu kepada Joko. "Sepertinya masalah ini bisa dipecahkan dengan algoritma apriori deh" Ucap saya. Lalu Joko meminta saya untuk memecahkan masalahnya tersebut.

**Menginstal Library**

In [1]:
#Pertama-tama kita import beberapa library ke dalam program ini seperti pandas, numpy, dan apyori
!pip install apyori

import pandas as pd
import numpy as np

from apyori import apriori

  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5974 sha256=afa6020bcfdc3bdd6026a25d2b8779ca917f8df24ab1556561eca1be217a19ff
  Stored in directory: /root/.cache/pip/wheels/cb/f6/e1/57973c631d27efd1a2f375bd6a83b2a616c4021f24aab84080
Successfully built apyori


**Meload Dataset**

In [5]:
# Meload dataset yang berisikan ID Transaksi dan Items 
df= pd.read_excel('/content/drive/MyDrive/Items.xlsx')


**Cleaning Data**

In [8]:
# Menghapus kolom ID Transaksi
dropped = ['ID Transaksi']
df.drop(columns=dropped, inplace=True)
df.head()

Items
0  Broccoli, Green Pappers, Corns, Broccoli
1        Asparagus, Squash, Corns, Tomatoes
2            Corns, Tomatoes, Beans, Squash
3     Green Pappers, Corns, Tomatoes, Beans
4  Broccoli, Corns, Tomatoes, Green Pappers

In [9]:
# Melihat shape dari tabel, yang sebelumnya 2 kolom 11 baris
df.shape

(11, 1)

In [10]:
# Membuat variabel data
data = df

**Membuat Data Training**

In [11]:
records = []
for i in range(data.shape[0]):
    records.append([str(data.values[i,j]).split(',') for j in range(data.shape[1])])

trx = [[] for trx in range(len(records))]
for i in range(len(records)):
    for j in records[i][0]:
        trx[i].append(j)
trx

[['Broccoli', ' Green Pappers', ' Corns', ' Broccoli'],
 ['Asparagus', ' Squash', ' Corns', ' Tomatoes'],
 ['Corns', ' Tomatoes', ' Beans', ' Squash'],
 ['Green Pappers', ' Corns', ' Tomatoes', ' Beans'],
 ['Broccoli', ' Corns', ' Tomatoes', ' Green Pappers'],
 ['Squash', ' Asparagus', ' Beans', ' Tomatoes'],
 ['Tomatoes', ' Corns', ' Green Pappers'],
 ['Broccoli', ' Tomatoes', ' Green Pappers', ' Beans'],
 ['Squash', ' Asparagus', ' Beans', ' Broccoli', ' Green Pappers'],
 ['Beans', ' Corns', ' Squash', ' Asparagus', ' Beans'],
 ['Squash', ' Tomatoes', ' Broccoli']]

**Melakukan Perhitungan Apriori**

In [19]:
# Menggunakan fungsi apriori untuk membuat asosiasi
association_rules = apriori (trx)
# Membuat list untuk menampilkan hasil
association_results = association_rules

**Menampilkan Hasil Perhitungan**

In [20]:
# Menampilkan hasil dari penerapan assosiation rule
pd.set_option('max_colwidth', 1000)
Result=pd.DataFrame(columns=['Rule','Support','Confidence'])
for item in association_results:
    pair = item[2]
    for i in pair:
        items = str([x for x in i[0]])
        if i[3]!=1:
            Result=Result.append({
                'Rule':str([x for x in i[0]])+ " -> " +str([x for x in i[1]]),
                'Support':str(round(item[1]*100,2))+'%',
                'Confidence':str(round(i[2] *100,2))+'%'
                },ignore_index=True)
Result

Rule Support Confidence
0                      [' Asparagus'] -> [' Beans']  27.27%     100.0%
1                      [' Beans'] -> [' Asparagus']  27.27%      50.0%
2                      [' Asparagus'] -> ['Squash']  18.18%     66.67%
3                      ['Squash'] -> [' Asparagus']  18.18%     66.67%
4                          [' Beans'] -> [' Corns']  18.18%     33.33%
5                          [' Corns'] -> [' Beans']  18.18%     33.33%
6                  [' Beans'] -> [' Green Pappers']  18.18%     33.33%
7                  [' Green Pappers'] -> [' Beans']  18.18%      40.0%
8                         [' Beans'] -> [' Squash']  18.18%     33.33%
9                         [' Squash'] -> [' Beans']  18.18%     66.67%
10                      [' Beans'] -> [' Tomatoes']  36.36%     66.67%
11                      [' Tomatoes'] -> [' Beans']  36.36%     57.14%
12                         [' Beans'] -> ['Squash']  18.18%     33.33%
13                         ['Squash'] -> [' Beans']  18.18%     66.67%
14              [' Broccoli'] -> [' Green Pappers']  18.18%     66.67%
15              [' Green Pappers'] -> [' Broccoli']  18.18%      40.0%
16                      [' Broccoli'] -> ['Squash']  18.18%     66.67%
17                      ['Squash'] -> [' Broccoli']  18.18%     66.67%
18                 [' Corns'] -> [' Green Pappers']  27.27%      50.0%
19                 [' Green Pappers'] -> [' Corns']  27.27%      60.0%
20                        [' Corns'] -> [' Squash']  18.18%     33.33%
21                        [' Squash'] -> [' Corns']  18.18%     66.67%
22                      [' Corns'] -> [' Tomatoes']  27.27%      50.0%
23                      [' Tomatoes'] -> [' Corns']  27.27%     42.86%
24                       [' Corns'] -> ['Broccoli']  18.18%     33.33%
25                       ['Broccoli'] -> [' Corns']  18.18%     66.67%
26              [' Green Pappers'] -> [' Tomatoes']  18.18%      40.0%
27              [' Tomatoes'] -> [' Green Pappers']  18.18%     28.57%
28               [' Green Pappers'] -> ['Broccoli']  27.27%      60.0%
29               ['Broccoli'] -> [' Green Pappers']  27.27%     100.0%
30                     [' Squash'] -> [' Tomatoes']  18.18%     66.67%
31                     [' Tomatoes'] -> [' Squash']  18.18%     28.57%
32                    [' Tomatoes'] -> ['Broccoli']  18.18%     28.57%
33                    ['Broccoli'] -> [' Tomatoes']  18.18%     66.67%
34                      [' Tomatoes'] -> ['Squash']  18.18%     28.57%
35                      ['Squash'] -> [' Tomatoes']  18.18%     66.67%
36           [' Asparagus'] -> [' Beans', 'Squash']  18.18%     66.67%
37           [' Beans'] -> [' Asparagus', 'Squash']  18.18%     33.33%
38           ['Squash'] -> [' Asparagus', ' Beans']  18.18%     66.67%
39           [' Asparagus', ' Beans'] -> ['Squash']  18.18%     66.67%
40           [' Asparagus', 'Squash'] -> [' Beans']  18.18%     100.0%
41           [' Beans', 'Squash'] -> [' Asparagus']  18.18%     100.0%
42     [' Corns'] -> [' Green Pappers', 'Broccoli']  18.18%     33.33%
43     [' Green Pappers'] -> [' Corns', 'Broccoli']  18.18%      40.0%
44     ['Broccoli'] -> [' Corns', ' Green Pappers']  18.18%     66.67%
45     [' Corns', ' Green Pappers'] -> ['Broccoli']  18.18%     66.67%
46     [' Corns', 'Broccoli'] -> [' Green Pappers']  18.18%     100.0%
47     [' Green Pappers', 'Broccoli'] -> [' Corns']  18.18%     66.67%
48  [' Green Pappers'] -> [' Tomatoes', 'Broccoli']  18.18%      40.0%
49  [' Tomatoes'] -> [' Green Pappers', 'Broccoli']  18.18%     28.57%
50  ['Broccoli'] -> [' Green Pappers', ' Tomatoes']  18.18%     66.67%
51  [' Green Pappers', ' Tomatoes'] -> ['Broccoli']  18.18%     100.0%
52  [' Green Pappers', 'Broccoli'] -> [' Tomatoes']  18.18%     66.67%
53  ['Broccoli', ' Tomatoes'] -> [' Green Pappers']  18.18%     100.0%

Dari hasil diatas kita dapat menyimpulkan barang mana saja yang bisa dijual Rudi pada Flash-Sale yang mempunyai nilai confidence tinggi:

*  [' Asparagus'] -> [' Beans']	      27.27%	100.0%
*  ['Broccoli'] -> [' Green Pappers']	27.27%	100.0%
*  [' Asparagus', 'Squash'] -> [' Beans'] 18.18%	 100.0%
*  [' Beans', 'Squash'] -> [' Asparagus']	      18.18%	100.0%

*  [' Corns', 'Broccoli'] -> [' Green Pappers']	18.18%	100.0%
*  [' Green Pappers', ' Tomatoes'] -> ['Broccoli']	18.18%	100.0%

*   ['Broccoli', 'Tomatoes'] -> [' Green Pappers']18.18%	100.0%

Dengan hasil ini kita dapat menyelesaikan masalah yang dialami Rudi